File: dekaf-core-test-db/src/main/java/org/jetbrains/dekaf/core/UnknownDBTestHelper.java
package org.jetbrains.dekaf.core;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.dekaf.exceptions.UnknownDBException;
import org.jetbrains.dekaf.sql.Scriptum;

import java.sql.*;
import java.util.ArrayList;



/**
 * @author Leonid Bushuev
 **/
public class UnknownDBTestHelper extends BaseTestHelper<DBFacade> {

  private boolean initialized;
  private boolean isOracle, isDB2;

  private String fromSingleRowTable = "";

  public UnknownDBTestHelper(@NotNull final DBFacade db) {
    super(db, Scriptum.of(UnknownDBTestHelper.class));
    if (db.isConnected()) initVariables();
  }

  private void initVariablesIfNeeded() {
    if (!initialized) {
      assert db.isConnected() : "Expected that is connected to DB";
      initVariables();
    }
  }

  private void initVariables() {
    ConnectionInfo info = db.getConnectionInfo();
    isOracle = info.rdbmsName.contains("ORACLE");
    isDB2 = info.rdbmsName.startsWith("DB2");
    fromSingleRowTable = isOracle ? " from dual" : isDB2 ? " from sysibm.sysdummy1" : "";
    initialized = true;
  }

  @Override
  public void prepareX1() {
    performCommand("create or replace view X1 as select 1 as X" + fromSingleRowTable);
  }

  @Override
  public void prepareX1000() {
    performCommand(scriptum, "X10");
    performCommand(scriptum, "X1000");
  }

  @Override
  public void prepareX1000000() {
    prepareX1000();
    performCommand(scriptum, "X1000000");
  }

  @NotNull
  @Override
  public String fromSingleRowTable() {
    initVariablesIfNeeded();
    return fromSingleRowTable;
  }


  @Override
  public void ensureNoTableOrView(final String... names) {
    final String[] tableTypes = new String[] {"TABLE","VIEW"};

    final ConnectionInfo connectionInfo = db.getConnectionInfo();
    if (connectionInfo.databaseName == null)
      throw new IllegalStateException("Cannot clean schema when the database name is unknown");
    if (connectionInfo.schemaName == null)
      throw new IllegalStateException("Cannot clean schema when the schema name is unknown");

    db.inSession(new InSessionNoResult() {
      @Override
      public void run(@NotNull final DBSession session) {

        for (String name : names) {
          zapTables(session, connectionInfo, name, tableTypes);
        }

      }
    });
  }

  @Override
  public void zapSchema() {
    final ConnectionInfo connectionInfo = db.getConnectionInfo();
    if (connectionInfo.databaseName == null)
      throw new IllegalStateException("Cannot clean schema when the database name is unknown");
    if (connectionInfo.schemaName == null)
      throw new IllegalStateException("Cannot clean schema when the schema name is unknown");

    db.inSession(new InSessionNoResult() {
      @Override
      public void run(@NotNull final DBSession session) {

        zapTables(session, connectionInfo, "%", null);

      }
    });
  }



  @SuppressWarnings("ThrowFromFinallyBlock")
  private void zapTables(final DBSession session,
                         final ConnectionInfo connectionInfo,
                         final String tableNameMask, final String[] tableTypes) {
    DatabaseMetaData md =
        session.getSpecificService(DatabaseMetaData.class,
                                   ImplementationAccessibleService.Names.JDBC_METADATA);
    if (md == null) throw new RuntimeException("Unable to obtain metadata from unknown database");

    // obtaining tables and list of commands to execute
    final ArrayList<String> commands = new ArrayList<String>();
    try {
      ResultSet tables = md.getTables(connectionInfo.databaseName,
                                      connectionInfo.schemaName,
                                      tableNameMask,
                                      tableTypes);
      try {
        while (tables.next()) {
          String tableName = tables.getString(3);
          String tableType = tables.getString(4);
          if (tableName == null || tableName.isEmpty() || tableType == null || tableType.isEmpty()) continue;
          tableType = tableType.toLowerCase();
          if (tableType.contains("system")) continue;
          String command = "drop " + tableType + ' ' + '"' + tableName + '"';
          commands.add(command);
        }
      }
      finally {
        tables.close();
      }
    }
    catch (SQLException e) {
      throw new UnknownDBException("Unable to list tables of unknown schema: " + e.getMessage(), e, null);
              // TODO try to recognize the error
    }

    // execute commands
    final ArrayList<String> errors = new ArrayList<String>(0);
    final Connection connection =
        session.getSpecificService(Connection.class,
                                   ImplementationAccessibleService.Names.JDBC_CONNECTION);
    assert connection != null;
    try {
      Statement stmt = connection.createStatement();
      try {
        for (String command : commands) {
          try {
            stmt.execute(command);
          }
          catch (SQLException e) {
            String errMessage = "Failed to " + command + ": " + e.getMessage();
            errors.add(errMessage);
          }
        }
      }
      finally {
        stmt.close();
      }
    }
    catch (SQLException e) {
      throw new UnknownDBException("Unable to list tables of unknown schema: " + e.getMessage(), e, null);
      // TODO try to recognize the error
    }

    if (!errors.isEmpty()) {
      StringBuffer b = new StringBuffer(errors.size()*100);
      b.append(errors.size()).append(" errors occurred when attempted to clean the schema:\n");
      for (String error : errors) b.append('\t').append(error).append('\n');
      throw new RuntimeException(b.toString());
    }
  }
}

File: dev/dekaf-test-db/src/core/UnknownDBTestHelper.kt
package org.jetbrains.dekaf.core

import org.jetbrains.dekaf.exceptions.UnknownDBException
import java.sql.Connection
import java.sql.DatabaseMetaData
import java.sql.SQLException
import java.util.*


/**
 * A generic test helper for other (unsupported) RDBMS.
 *
 * @author Leonid Bushuev
 */
class UnknownDBTestHelper(db: DBFacade) : BaseTestHelper(db, UnknownDBTestHelper::class) {

    private var initialized: Boolean = false
    private var isOracle: Boolean = false
    private var isDB2: Boolean = false
    private var isHsql: Boolean = false
    private var isDerby: Boolean = false

    private var fromSingleRowTable = ""

    init {
        if (db.isConnected) initVariables()
    }

    private fun initVariablesIfNeeded() {
        if (!initialized) {
            assert(db.isConnected) { "Expected that is connected to DB" }
            initVariables()
        }
    }

    private fun initVariables() {
        val info = db.connectionInfo
        isOracle = info.rdbmsName.startsWith("Oracle")
        isDB2 = info.rdbmsName.startsWith("DB2")
        isHsql = info.rdbmsName.startsWith("HSQL")
        isDerby = info.rdbmsName.contains("Derby")

        fromSingleRowTable = when {
            isOracle         -> " from dual"
            isDB2 || isDerby -> " from sysibm.sysdummy1"
            isHsql           -> " from information_schema.schemata limit 1"
            else             -> ""
        }
        
        initialized = true
    }

    override fun prepareX1() {
        performCommand("create or replace view X1 as select 1 as X" + fromSingleRowTable)
    }

    override fun prepareX1000() {
        performCommand(scriptum, "X10")
        performCommand(scriptum, "X1000")
    }

    override fun prepareX1000000() {
        prepareX1000()
        performCommand(scriptum, "X1000000")
    }

    override fun fromSingleRowTable(): String {
        initVariablesIfNeeded()
        return fromSingleRowTable
    }


    override fun ensureNoTableOrView(vararg names: String) {
        val tableTypes = arrayOf("TABLE", "VIEW")

        val connectionInfo = db.connectionInfo
        if (connectionInfo.databaseName == null)
            throw IllegalStateException("Cannot clean schema when the database name is unknown")
        if (connectionInfo.schemaName == null)
            throw IllegalStateException("Cannot clean schema when the schema name is unknown")

        db.inSessionDo { session ->

            for (name in names) {
                zapTables(session, connectionInfo, name, tableTypes)
            }

        }
    }

    override fun zapSchema() {
        val connectionInfo = db.connectionInfo
        if (connectionInfo.databaseName == null)
            throw IllegalStateException("Cannot clean schema when the database name is unknown")
        if (connectionInfo.schemaName == null)
            throw IllegalStateException("Cannot clean schema when the schema name is unknown")

        db.inSessionDo { session ->

            zapTables(session, connectionInfo, "%", null)

        }
    }


    private fun zapTables(session: DBSession,
                          connectionInfo: ConnectionInfo,
                          tableNameMask: String, tableTypes: Array<String>?) {
        val md = session.getSpecificService(DatabaseMetaData::class.java,
                                            ImplementationAccessibleService.Names.JDBC_METADATA) ?: throw RuntimeException("Unable to obtain metadata from unknown database")

        // obtaining tables and list of commands to execute
        val commands = ArrayList<String>()
        try {
            val tables = md.getTables(connectionInfo.databaseName,
                                      connectionInfo.schemaName,
                                      tableNameMask,
                                      tableTypes)
            try {
                while (tables.next()) {
                    val tableName = tables.getString(3)
                    var tableType: String? = tables.getString(4)
                    if (tableName == null || tableName.isEmpty() || tableType == null || tableType.isEmpty()) continue
                    tableType = tableType.toLowerCase()
                    if (tableType.contains("system")) continue
                    val command = "drop $tableType \"$tableName\""
                    commands.add(command)
                }
            }
            finally {
                tables.close()
            }
        }
        catch (e: SQLException) {
            throw UnknownDBException("Unable to list tables of unknown schema: " + e.message, e, null)
            // TODO try to recognize the error
        }

        // execute commands
        val errors = ArrayList<String>(0)
        val connection = session.getSpecificService(Connection::class.java,
                                                    ImplementationAccessibleService.Names.JDBC_CONNECTION)!!
        try {
            val stmt = connection.createStatement()
            try {
                for (command in commands) {
                    try {
                        stmt.execute(command)
                    }
                    catch (e: SQLException) {
                        val errMessage = "Failed to " + command + ": " + e.message
                        errors.add(errMessage)
                    }

                }
            }
            finally {
                stmt.close()
            }
        }
        catch (e: SQLException) {
            throw UnknownDBException("Unable to list tables of unknown schema: " + e.message, e, null)
            // TODO try to recognize the error
        }

        if (!errors.isEmpty()) {
            val b = StringBuffer(errors.size * 100)
            b.append(errors.size).append(" errors occurred when attempted to clean the schema:\n")
            for (error in errors) b.append('\t').append(error).append('\n')
            throw RuntimeException(b.toString())
        }
    }
}
